### PR TITLE
Slapcraft Framework

### PR DESCRIPTION
**Pull request by Azarak**
## About The Pull Request
(This PR also adds some reagent holder helpers cause I needed to use such ones, and they were long overdue for existance)
(I will convert more recipes in other PRs for the sake of atomization and ease of review)

Adds a Slapcraft Framework, which is a system that easily allows implementation of crafting recipes which rely on slapping items against one another. There is a button next to a crafting menu that opens a handbook with recipes for it.
Currently this comes in with only 2 recipes, but more can be easily added. (IED and Flamethrower are the ones)
There is a HUD element which opens a Handbook with all the recipes inside.
The steps inside the handbook can link to another recipe for easy navigation across crafting components. (Currently none utilize this)

This system utilizes datums with some robust features and virtual procs to support about any behaviour in crafting that you might desire. I have coded in base steps for using stack items, for using tools, using weapons, using reagents, or placing a reagent container with a required amount of reagents inside.

You can examine an assembly of a crafted item in progress to know what steps have been done and how to do the next one.
You can also disassemble an unfinished assembly by using it in hand.

This system has user feedback in form of notice message aswell as sound effects.
Each step by default takes 2 seconds currently.

This system is to eventually replace the crafting menu. 

The recipes have support to have their steps be performed in order, (step by step) or first - rest whatever order then last, (first and last) or first then any other step. (freeform)

The steps have support to be optional, but optional steps cannot be first or last steps of any recipe, and they cannot be used in freeform order recipes.

Items which are required for the first step of a recipe show a hint about the craft in examine.
dreamseeker_2022-06-12_01-09-17

This system even handles collisions by utilizing a radial menu: (test recipes showcased)
dreamseeker_2022-06-01_17-26-46

Crafting a flamethrower:
https://cdn.discordapp.com/attachments/800146077047062558/981791330554495067/2022-06-02-07-24-23.webm
Handbook:

Codewise it's as easy as this to make a recipe for something complex
sublime_text_2022-06-02_08-29-14

## Why It's Good For The Game
Immersive, satisfying, snappy.

## Changelog
:cl:
add: Adds a Slapcraft Framework, which is a system that easily allows implementation of crafting recipes which rely on slapping items against one another. There is a button next to a crafting menu that opens a handbook with recipes for it.
del: Removes crafting menu IED and Flamethrower recipes
del: Removes hardcoded Flamethrower slapcrafting.
add: Adds IED and Flamethrower slapcrafting recipes, which can be found in the handbook.
/:cl:

### CODE REVIEW

**Comment by Azarak** [comment]
Ready to be reviewed but I will add a couple more features.

**Comment by Azarak** [comment]
Actually, I want to get another PR with abstract types merged first so I can use that in this PR

File: code/_onclick/hud/_defines.dm [hunk 1]
```diff
@@ -57,6 +57,7 @@
 #define ui_zonesel "EAST-1:28,SOUTH:5"
 #define ui_acti_alt "EAST-1:28,SOUTH:5" //alternative intent switcher for when the interface is hidden (F12)
 #define ui_crafting "EAST-4:22,SOUTH:5"
+#define ui_slapcraft "EAST-4:22,SOUTH:5"
```
Comment (atakiya) [inline code/_onclick/hud/_defines.dm:60]: Would this not conflict with `ui_crafting` on the same coordinates?

Comment (Azarak) [reply code/_onclick/hud/_defines.dm:60]: Its icon is moved by an appropriate offset. It's good. It's the same way as other buttons.

File: code/modules/slapcraft/recipes/weapons/flamethrower.dm [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+		/datum/slapcraft_step/stack/flamethrower_rod,
+		/datum/slapcraft_step/tool/screwdriver/flamethrower
+		)
+	result_type = /obj/item/flamethrower
+
+/datum/slapcraft_step/flamethrower_welder
+	desc = "Start with normal sized welding tool."
+	finished_desc = "A welding tool has been added."
+	item_types = list(/obj/item/weldingtool)
+	blacklist_item_types = list(/obj/item/weldingtool/mini, /obj/item/weldingtool/largetank)
```
[the remark this answers is not in the document]
Comment (Azarak) [reply code/modules/slapcraft/recipes/weapons/flamethrower.dm:17]: Oh yeah totally, I must have forgot.

**Review by atakiya — APPROVED** [review]

**Pull request merged**